src/github/utils.ts

```diff
@@ -12,21 +12,89 @@ import { parseDiffHunk, DiffHunk } from '../common/diffHunk';
 import * as Common from '../common/timelineEvent';
 import * as GraphQL from './graphql';
 import { Resource } from '../common/resources';
+import { PullRequestModel } from './pullRequestModel';
+import { getEditCommand, getDeleteCommand, getCommentThreadCommands } from './commands';
+import { PRNode } from '../view/treeNodes/pullRequestNode';
+import { ReviewDocumentCommentProvider } from '../view/reviewDocumentCommentProvider';
+import { uniqBy } from '../common/utils';
+
+export interface CommentHandler {
+	commentController?: vscode.CommentController;
+	startReview(thread: vscode.CommentThread): Promise<void>;
+	finishReview(thread: vscode.CommentThread): Promise<void>;
+	deleteReview(): Promise<void>;
+	createOrReplyComment(thread: vscode.CommentThread): Promise<void>;
+	editComment(thread: vscode.CommentThread, comment: vscode.Comment): Promise<void>;
+	deleteComment(thread: vscode.CommentThread, comment: vscode.Comment): Promise<void>;
+}
 
-export function convertToVSCodeComment(comment: Comment, command?: vscode.Command): vscode.Comment {
-	return {
+export function convertToVSCodeComment(comment: Comment, command: vscode.Command | undefined): vscode.Comment & { _rawComment: Comment } {
+	let vscodeComment: vscode.Comment & { _rawComment: Comment } = {
+		_rawComment: comment,
 		commentId: comment.id.toString(),
 		body: new vscode.MarkdownString(comment.body),
-		command: command,
+		selectCommand: command,
 		userName: comment.user!.login,
-		gravatar: comment.user!.avatarUrl,
-		canEdit: comment.canEdit,
-		canDelete: comment.canDelete,
-		isDraft: !!comment.isDraft,
+		userIconPath: vscode.Uri.parse(comment.user!.avatarUrl),
+		label: !!comment.isDraft ? 'Pending' : undefined,
 		commentReactions: comment.reactions ? comment.reactions.map(reaction => {
 			return { label: reaction.label, hasReacted: reaction.viewerHasReacted, count: reaction.count, iconPath: reaction.icon };
 		}) : []
 	};
+
+	return vscodeComment;
+}
+
+export function createVSCodeCommentThread(thread: vscode.CommentThread, commentController: vscode.CommentController, pullRequestModel: PullRequestModel, inDraftMode: boolean, node: PRNode | ReviewDocumentCommentProvider) {
+	let vscodeThread = commentController.createCommentThread(
+		thread.threadId,
+		thread.resource,
+		thread.range!,
+		[]
+	);
+
+	thread.comments.forEach(comment => {

```

Comment:
could move this above and then call createCommentThread with vscodeThread.comments so there's no update event

Reply:
comment relies on `thread` as it's one argument for `edit` and `delete` command.

Reply:
Here I put `comments` in ctor and then update `comments` again with proper edit/delete commands, the reason is if we create comment thread with empty comments, there are chances that VS Code focuses in the comment editor.